### DISCUSSION TITLE
General questions

### DISCUSSION THREAD

**Discussion by OriolAbril** [felix-hilden/sphinx-codeautolink#44] [Ideas]
Hi! Thanks for the library, it looks very promising and I can't wait to try it. I would like to use the extension with [ArviZ](https://arviz-devs.github.io/arviz/index.html) docs. So far I haven't managed to get it to work so I'll try to go over my experience here so that you can hopefully guide me into opening proper issues for bug reports (if any) and enhancements as well as some PRs to improve documentation of the extension?

The first thing is that it took me a while to see if the extension would work at all in our case, as we don't really have any `.. code::` blocks, we use matplotlib's, ipython's or bokeh's directives to get the code to autoexecute and include the output, as well as jupyter notebooks parsed with myst_parser. All three directives seem to also use `highlight-python notranslate` so it looks like they should be compatible out of the box and I did try using the matplotlib example on your homepage to use the plot directive and worked. But it might be worth it to explain what are the actual requirements for this to work or include some examples in the docs, perhaps with ipython and matplotlib ones that are quite popular to show they work too.

myst seems to use `highlight-ipython3 notranslate` which I'm not sure how easy it would be to cover too. This page for example is generated from a jupyter notebook: https://arviz-devs.github.io/arviz/getting_started/Introduction.html.

I naively added the `highlight-ipython3 notranslate` to https://github.com/felix-hilden/sphinx-codeautolink/blob/master/src/sphinx_codeautolink/extension/block.py#L193 to see what happened and added a jupyter notebook file to the example docs in my fork. The code in the notebook was not hyperlinked, but at least it did not break the rendering nor links for the rest of the pages, which for now is good enough.

I eventually moved to try it on in ArviZ docs, installed the extension, added it to conf.py and set `codeautolink_autodoc_inject = False` as we don't want the auto tables there, they would quickly become quite large and we'd rather users focus on the examples in the docstring for quick reference or on the dedicated plotting, labeling... guides for specific tasks which are also useful even if the exact function is not being used as many arguments are common between most plotting functions for example. But I still haven't managed to get the docs to build, I get the following error:

```
Extension error (sphinx_codeautolink.extension):
Handler <bound method SphinxCodeAutoLink.generate_backref_tables of <sphinx_codeautolink.extension.SphinxCodeAutoLink object at 0x7f8d6a43b2e0>> for event 'doctree-resolved' threw an exception (exception: 'builtin_function_or_method' object has no attribute '__annotations__')
```

I am trying to build only part of the docs to see if I can identify what triggers that error, but so far all the subsets I have tried end up just like this. I was also wondering if this step is always necessary or if it's only necessary when planning to generate example tables either automatically or with the directive. In our case, even a hacky workaround that completely disables the feature and only adds links to code would be amazing, not sure if it's possible. 

I also removed autodoc in one of the experiments when trying to remove the api section completely from the equation which triggered a whole different error. I didn't completely understood the error nor why was it triggered by autodoc being missing, but I think the extension would be also extremely useful in documentation that doesn't use autodoc at all. I am thinking for example about collections of tutorials and case studies like https://pymc-examples.readthedocs.io/en/latest/ that thanks to the integration with intersphinx could easily link from case studies to the api docs in the main documentation website.

**Comment by felix-hilden** [comment]
Hi and thanks! I moved this to a discussion to hopefully triage the individual issues a bit better. I'll try to go over it and maybe we'll get to a resolution 👍

#### Custom code blocks
Firstly, I was not aware of any third-party libraries using their own directives as code blocks, so it's no wonder that they don't work 😅 So we should figure out how the directives work. Different highlights don't really matter as long as the code is valid Python and we can ignore the violating parts similar to doctest prefixes. The CSS class `highlight-python notranslate` is only used to locate the HTML to insert links, not to parse the source. The classes will eventually be added there, but having extra candidates doesn't break anything. So yeah, if it's reasonably possibly doable, I'm very much in favor of supporting more block types!

#### Annotations error
Now with respect to your `__annotations__` error, that could very well be a bug in our logic! Without any indication as to what breaks it it's hard to tell though. Could you try to narrow it down by building some dummy documentation? And what do you mean by "a part of the documentation"? This error occurs when following type hints in the documented library to link more complex access like `lib.foo().attr` to `lib.Bar.attr`.

#### Disregarding backreference tables
Implementing a hacky version of this extension entirely outside the build process is possible by just reading HTML and Sphinx's result object inventory, but I'm quite opposed to that, which I'm sure you'll understand.

#### Autodoc-free documentation
I haven't considered documentation without autodoc, but the use case you presented is solid. If you've specified an intersphinx entry for the library you're trying to link to, I see no reason for it not to work. Obviously it didn't though, so I'm keen on seeing the error that was raised!

Hopefully this clears things up a bit! Let's discuss more and I'll open up more focused issues as we go 🤗

**Reply by OriolAbril** [reply]
Thanks! I didn't realize discussions were open, sorry about that, this is a much better place. I'll try to answer in different "answers"/"posts" so we can then if necessary continue in comments and focus a bit on each point independently.

**Comment by OriolAbril** [comment]
### Custom code blocks

I opened #45 which should serve as an example on how matplotlib's custom plot directive already seems to integrate with the extension seamlessly. You should only need to generate the docs from the PR locally or to set up [rtd PR previews](https://docs.readthedocs.io/en/stable/pull-requests.html). Here are the main third party code directives I am aware of:

* https://matplotlib.org/stable/api/sphinxext_plot_directive_api.html?highlight=plot_directive#module-matplotlib.sphinxext.plot_directive
* https://ipython.readthedocs.io/en/stable/sphinxext.html
* https://docs.bokeh.org/en/latest/docs/reference/sphinxext.html#module-bokeh.sphinxext.bokeh_plot
* MyST/jupyter notebooks. Not sure which is the best resource to share as an explanation for that. AFAIK, the `code-cell` directive
  is used either to parse the code cells in jupyter notebook or used as a "regular" directive within markdown files. Maybe https://myst-nb.readthedocs.io/en/latest/use/markdown.html is a good starting point.
* https://jupyter-sphinx.readthedocs.io/en/latest/

**Reply by felix-hilden** [reply]
Some background to the Sphinx process since you seem interested: the documents are parsed to node trees, similar to AST. First they are preprocessed, where the matplotlib seems to convert the plot directive to a regular code block, then ran through various transformations and finally translated to HTML. So before the translation we catch all the interesting blocks and parse the code within. Adding new types of nodes is a matter of:

- Finding out the block type and its syntax
- Extracting the valid Python code
- Saving the original representation for matching
- Matching the result HTML with a preset CSS class and injecting links

Thanks for the list, I'll see what the directives are like and experiment with them!

**Reply by OriolAbril** [reply]
For myst type code cells, it look like adding `ipython3` in https://github.com/felix-hilden/sphinx-codeautolink/blob/master/src/sphinx_codeautolink/extension/block.py#L123 and https://github.com/felix-hilden/sphinx-codeautolink/blob/master/src/sphinx_codeautolink/extension/block.py#L195 statements will get the blocks to be parsed and get links working. I did a very quick test on that and only changing these two lines gets links placed in those cells, but only if we restrict ourselves to python syntax only. The main issue whichh will also be relevant for ipython and jupyter sphinx extension is that in such cases the code needs to be preprocessed (a bit like with pycon) to remove all magic and other special ipython syntax that is not valid python. This reference might help with that: https://ipython.readthedocs.io/en/stable/config/inputtransforms.html

**Reply by felix-hilden** [reply]
I added support just now, let me know if anything doesn't work still!

**Reply by OriolAbril** [reply]
Thanks! I'll probably need some time to get the ipython cleaner working, but that should make it easy. I do have one comment which I am not sure if it's some inconsistency (at some level, could as well be on our side, matplotlib/bokeh directive, sphinx itself) or something done by design. It looks like the output classes are not necessarily the same as the code blocks nor are always the same for the same directive. 

This page for example: https://arviz-devs.github.io/arviz/examples/plot_autocorr.html uses matplotlilb's plot directive and bokeh's plot directive to generate the code cells and images, and the output code has class `highlight-default notranslate` for both of them, However,  https://arviz-devs.github.io/arviz/api/generated/arviz.plot_elpd.html which uses the same two directives generates outputs with classes `doctest highlight-default notranslate` for matplotlib (I guess it is the same if ignoring the doctest) and `highlight-python notranslate` for bokeh one.

IIUC, with the changes you have done now, both code blocks in the docstring would be parsed correctly and had links added to them, but the ones in the examples would not. Is that right?

**Reply by felix-hilden** [reply]
Yikes, if this is indeed the case then that'd be a problem. So this is only the first step then! We'll also introduce a config value for extending the searched CSS classes. I'll create an issue.

**Reply by felix-hilden** [reply]
#54

**Reply by OriolAbril** [reply]
It turned out to be relatively easy to find the ipython cleaner function, still have not tested it too thoroughly on magic or shell commands but it seems to work for now quite well. Here is the PR to add it to ArviZ docs, very little code added: https://github.com/arviz-devs/arviz/pull/1923

There are some issues in the preview I don't quite understand though, I'll list them here quickly and open issues for those that are issues and not me messing up the config. We do get links in code cells generated from jupyter notebooks so it looks very promising.

* The intersphinx links are not working. If you take a look at https://arviz--1923.org.readthedocs.build/en/1923/user_guide/pymc3_refitting_xr_lik.html for example, you'll see that links to arviz functions work, but the links to numpy, matplotlib or to `xr.apply_ufunc` have an incorrect link that is kind of a mix between arviz and their real html page. These are not working locally either but there they get a `../` prepended, so it looks like when adding the links, there is some processing gone wrong that
* From what I understand, all code blocks from a given file should be parsed together so that imports in the first cell are taken into account in following code cells. This does not seem to be the case :thinking:. From the same page for example, arviz and xarray functions get links generated thanks to the global header config, but numpy or matplotlib ones don't get links after the first cell.
* I'll look into it again as I am completely lost on this. Might be related to the default vs python css class thing. Links in docstrings (using the plot directive) don't seem to be working: https://arviz--1923.org.readthedocs.build/en/1923/api/generated/arviz.plot_compare.html. Even though they did work for me locally:
  ![image](https://user-images.githubusercontent.com/23738400/139824701-b82c3e11-c7a4-4dd6-945b-760a66649d37.png)

**Reply by felix-hilden** [reply]
Looking at the PR: you haven't enabled [concatenation](https://sphinx-codeautolink.readthedocs.io/en/stable/reference.html#directive-autolink-concat). That should fix at least some issues. For now, it isn't available as a global configuration.

**Reply by OriolAbril** [reply]
Oh, I see, so the `on` being the default means that using the directive without arguments enables that behaviour, not that the behaviour is enabled by default. Would it be possible to make a global config parameter to get concatenation of cells on a per file basis without needing the concat directive in every file?

**Reply by felix-hilden** [reply]
Yeah, I'll do it. I already took it away earlier for clarity, but since we allow a global preface there's not a convincing argument for explicitness in that regard anymore.

**Reply by OriolAbril** [reply]
I also tried it out on https://github.com/pymc-devs/pymc-examples/pull/239 with similar results. I have tried to debug but I don't really know what I'm doing. in https://pymc-examples--239.org.readthedocs.build/en/239/time_series/MvGaussianRandomWalk_demo.html for example only some of the modules configured in intersphinx have working links, and it doesn't look like it's an intersphinx config issue because in the inferencesection there are intersphinx references within the text pointing to arviz that work, but it's import is not highlighted. 

I thought it could be related to pymc and arviz being imported in the preface too, but the same happens after removing the preface.

Let me know if there are some tests that could help you understand what is happening.

**Reply by felix-hilden** [reply]
It *could* be a case of the documentation not including an entry for the module itself. If the autodoc use is very detailed, the top-level module entry might not even exist. But that would need a look at the object inventory.

**Reply by felix-hilden** [reply]
I added the global config in #55!

**Reply by OriolAbril** [reply]
Thanks!

I will check autodoc use, now that you say that, we do have quite detailed autodoc usage to prioritize conceptual groupings over file ones and might not have a highest level module entry.

**Reply by felix-hilden** [reply]
Nice, I also released 0.5.0 with the custom CSS classes option in #54, so hopefully it's easier to test now.

**Reply by felix-hilden** [reply]
Could you verify that the solution in #56 works for your more complex case as well? I might do 0.5.1 containing only that fix if it works fine.

**Reply by OriolAbril** [reply]
The intersphinx link is working, just tested locally. What doesn't seem to be working are the extra css classes. I have added the config param at https://github.com/arviz-devs/arviz/blob/main/doc/source/conf.py#L92 yet https://arviz-devs.github.io/arviz/api/generated/arviz.plot_autocorr.html doesn't have links in the code blocks, I'll try to generate a minimal example next week and open an issue. 

Everything else is working like a charm, thanks so much. We have already activated it on our published docs, and have for example https://arviz-devs.github.io/arviz/getting_started/Introduction.html with all relevant functions highlighted. For a page that comes from a jupyter notebook :tada:

**Reply by felix-hilden** [reply]
Glad to hear! And then not 😅 but we shall continue debugging. I'm away for the weekend though so next week with the MWE it is!

**Reply by felix-hilden** [reply]
Heya @OriolAbril! I had a quick look and when inspecting the HTML I noticed that at least for that particular example, the CSS class is `doctest highlight-default notranslate`. So after adding that in, the correct links are produced! Could you verify that this is indeed the case and perhaps fixes other problem cases as well? Maybe you have even more classes that need to be covered.

**Reply by OriolAbril** [reply]
I am overwhelmed right now and keep thinking I'll do it in 2 days, or next week but still can't do it then. I have a note to check this but it will probably take a while until I can work on docs.

**Reply by felix-hilden** [reply]
No worries! Everything seems to be working fine 👍 but I'm still interested to hear how it went once you get back to unpaid work 😄

**Reply by OriolAbril** [reply]
Trying it out at https://github.com/arviz-devs/arviz/pull/1956, doesn't seem to work so far. I have also realized that these other code blocks are only `highlight-default notranslate` and are not getting links added either: https://arviz--1956.org.readthedocs.build/en/1956/examples/plot_autocorr.html

**Reply by felix-hilden** [reply]
Thanks for reporting! I added some other comments to the PR, but I'm not sure what's happening here. At first I thought that the culprit is `sphinx-copybutton`, which modifies the innermost `pre` to have `<pre id="codecell0">` and adds a button after, but surely they shouldn't bother us. Did you upgrade? Warnings or errors weren't issued at least in the PR docs run. How about locally or with the newest version?

**Reply by OriolAbril** [reply]
just checked the latest build in the PR and it is using 0.8.0 which is the latest version right? Thanks for the PR review by the way :heart:

**Reply by felix-hilden** [reply]
You're quite welcome, and yeah it is! But the build still fails. By the way, should something special be added for `ipython` (not "3") blocks? I see you've commented something out and noted that linking doesn't work. Could it be that the cell is that type instead of `ipython3`? I'm still a bit shaky on exactly what happens in different IPython scenarios 😅

**Reply by OriolAbril** [reply]
There is also the [ipython directive](https://ipython.readthedocs.io/en/stable/sphinxext.html) which we use for examples without plots (for example https://arviz-devs.github.io/arviz/api/generated/arviz.compare.html) is that it's not only a matter of cleaning the leading In[], ..., and Out[] elements but also removing/commenting out the output because everhthing is added as a single code block.

As I commented I tried to comment output and clean up the input to valid python code which worked for the first step, but then it wasn't able to match sources to html output and actually add the links.

**Reply by felix-hilden** [reply]
Yeah, the original source must be identical to the HTML representation for matching to work. And it would need a transform function that only ignores lines instead of changing them. Looks like it doesn't really work with the basic transformer. Would you be interested in helping me sort it out? Seems like we should support that directive as well if we claim to support IPython.

If this isn't the problem, could you try out the current master so we are better able to see if any warnings are issued? We recently started using Sphinx logging for better visibility. `pip install git+https://github.com/felix-hilden/sphinx-codeautolink.git` I believe works.

**Reply by felix-hilden** [reply]
Seems somewhat difficult though, since they are passed to the interpreter and the line numbers and output may change 😮

**Reply by OriolAbril** [reply]
I can write down some examples of ipython directive usage to play around. Would you prefer a PR adding the examples to codeautolink docs? Or creating a minimal repo with it?

**Reply by OriolAbril** [reply]
I did try the development version in the last commit of my PR in case it helps with the `highlight-default` issue. Here are the sphinx logs: https://readthedocs.org/api/v2/build/15693463.txt

**Reply by felix-hilden** [reply]
Just some examples for now would be good! But indeed there are no warnings, so most likely the blocks aren't even parsed. Do you have any information on what the cell type is? Always `ipython3`?

**Reply by OriolAbril** [reply]
It's `ipython` instead of `ipython3`. Here is the permalink to what I tried in case it's helpful, I'll share some examples soon

**Reply by felix-hilden** [reply]
There's no link, did you mean to add one? And thanks! It seems quite odd that `ipython` block type is used for both straight code cells like in the problem case you presented and interpreter sessions with the `In[...] / Out[...]` prefixes like with the `ipython` directive.

**Reply by OriolAbril** [reply]
sorry, here is the link: https://github.com/arviz-devs/arviz/blob/ec789c8c51252113c0ce8fadc7f676def645b371/doc/source/conf.py#L89

**Reply by felix-hilden** [reply]
Oh I see, that can't really work because it returns the cleaned up version with comments for matching, right? The first return value must be the text representation of the code block in the result HTML, so most likely the original source unchanged, or in the case of `pycon`, `<BLANKLINE>` tags removed.

**Reply by OriolAbril** [reply]
ooohhh, then it could work with very little change. I tried to modularize the changes wrong. Could something like the code below work?

```
def ipython_directive_transform(source):
    """Ignore lines starting with 'In[]:' or '...:'."""
    lines = []
    for line in source.split("\n"):
        (line, num_subs) = re.subn(r"^\s*(In\s*\[[0-9]+\]|\.{3,})\:\s", "", line)
        if num_subs==1:
            lines.append(line)
        else:
            lines.append(f"# {line}")
    return source, TransformerManager().transform_cell("\n".join(lines))
```

**Reply by felix-hilden** [reply]
For sure! Or even `else: lines.append("")` like we have with `pycon`. The regex is somewhat complicated looking, at least for me, so that could be changed to something similar to [this logic](https://github.com/felix-hilden/sphinx-codeautolink/blob/master/src/sphinx_codeautolink/extension/block.py#L36) which also handles the case of output including three dots.

Do you have any experience with the directive vs. notebook `ipython` blocks to know why some have those `In[...]` prefixes and your blocks don't? I'm still somewhat worried about that, and the fact that the line numbers and output may change.

**Reply by OriolAbril** [reply]
notebook blocks are `ipython3` only I think, not `ipython`. I am not 100% sure how it works. It might be a myst decision to skip the numbers or a theme level one. What I do know is that with our current setting, notebook code cells are identified by the green left border which other code blocks don't have. 

I _think_ that nbsphinx/other themes show the cell number, but it's handled independently of the code, it's not like in the `ipython` directive where the `In` and `Out` things are part of the code block (even when the output is generated by sphinx executing the code)

**Reply by felix-hilden** [reply]
Thanks! I've created an issue for this: #91. Once more, both the autocorrelation and compare examples are using the `ipython` directive, even though autocorr doesn't have prefixes?

**Reply by OriolAbril** [reply]
No, compare uses the ipython direcrive, autocorr uses the matplotlib plot directive

**Reply by felix-hilden** [reply]
Oh, so these are completely separate issues then, and the autocorrelation should surely work with the currently existing machinery. Could it have something to do with generating the files? How do you do it?

**Reply by OriolAbril** [reply]
Not sure, we use the directive in the docstrings and then use autosummary to include them in the docs

**Reply by felix-hilden** [reply]
Right, autosummary doesn't generate RST files, so that might be the key. This needs to be tested, but my current guess is that either the blocks aren't analysed at all, or they are not saved or cannot be parsed in HTML because the files haven't come from concrete sources. I'm on it!

Or wait, does it? Well anyways I'll investigate.

**Reply by OriolAbril** [reply]
it can do, and in our case it does. We have written https://github.com/arviz-devs/arviz/blob/main/doc/source/api/plots.rst which when building the documentation writes all the object specific rst files in the `generated/` folder. I have never really checked what is actually the content of those files though

**Comment by OriolAbril** [comment]
### Annotations error
I started trying to get a minimal build that reproduces the error, but I didn't have much time and wasn't able to get anywhere, a couple cases were very simplified and worked, a couple others had different errors (like removing the whole api section and consequently not using autodoc). 

I will try to get a mwe and get back to you here.

**Reply by OriolAbril** [reply]
I tracked the error down to `np.random.normal` and similar calls, see https://github.com/felix-hilden/sphinx-codeautolink/pull/47 as a way to trigger it with two extra lines only in your intersphinx example.

**Reply by felix-hilden** [reply]
I had a look at `np.random.normal`. It's a method (bound to some underlying object) but despite that it doesn't have an `__annotations__` dict. Even unannotated functions and lambdas should have them though (as an empty dict), so something odd is going on there. Perhaps C bindings or something, further investigation needed. As to why Bokeh breaks the linking.. no idea still. The type hints shouldn't be followed, because there isn't any attribute access after the call.

**Comment by OriolAbril** [comment]
### Disregarding backreference tables
I completely understand, and my goal is to get it to work and to also be able to use the example tables, even if not automatically nor everywhere. As you can see from my hichlight-ipython3 approach, I am an avid sphinx user, but I have never written an extension nor I understand how they work nor how much of a challenge would supporting extra third party code directives be. A solution of that sort would be as a last resort which doesn't seem necessary after your thoughtful response

**Reply by felix-hilden** [reply]
I agree, and we can consider this resolved then!

**Comment by OriolAbril** [comment]
### Autodoc-free documentation
I will share the error as soon as possible. This I do know how to reproduce in a mwe so it should not take too much time. Thanks for considering this!

**Reply by felix-hilden** [reply]
Was the error "unknown event name: autodoc-process-docstring"?

**Reply by OriolAbril** [reply]
Yes!

**Reply by felix-hilden** [reply]
Alright, we need a way of disabling our autodoc integration conditionally. I'll see what can be done. As a last resort we could require autodoc (can be done in code if the user hasn't enabled it which is nice) even though the documentation doesn't use it.

**Reply by felix-hilden** [reply]
Seems that Sphinx doesn't have a good way of doing it, so we'll go with activating autodoc anyway: #48
